### ISSUE TITLE
multiple states with same turn number 

### ISSUE THREAD

**Issue by andrewgordstewart** [statechannels/monorepo#1842]
[web3torrent.Thu, 14 May 2020 02_21_26 GMT.f0abd39f4.0xEF99c5279436f7d015A8AB32505721543e9DAeC7.pino.log](https://github.com/statechannels/monorepo/files/4625686/web3torrent.Thu.14.May.2020.02_21_26.GMT.f0abd39f4.0xEF99c5279436f7d015A8AB32505721543e9DAeC7.pino.log)

Reproduced locally at https://github.com/statechannels/monorepo/pull/1841.

# Current status

## Web3torrent issue
Web3torrent had a [bug](https://github.com/statechannels/monorepo/commit/fce88d4e20329355a7008b4d1b40fccf53628375) that caused it to trigger multiple state updates concurrently, (via `channelClient.updateChannel`) for a single payment channel. This issue is resolved.

## Client issue
The behaviour of `channelClient.updateChannel` is [undetermined and undocumented.](https://github.com/statechannels/monorepo/issues/1853) See below for discussion.

## Wallet issue
The `application` workflow [updates the state](https://github.com/statechannels/monorepo/blob/master/packages/xstate-wallet/src/workflows/application.ts#L341-L354) in an **action**, and not in a **transaction**
- actions are _fire and forget_. Should it happen in an action? Or should it be a service?
- because it's not a **transaction**, when it asks the store to [sign the `newState`](https://github.com/statechannels/monorepo/blob/master/packages/xstate-wallet/src/workflows/application.ts#L352), the `newState` may now be out of date.

**Comment by andrewgordstewart** [comment]
Better log (with more informative store error): 
[web3torrent.Thu, 14 May 2020 02_42_14 GMT.f0abd39f4.0xC11C4a744747bEc56504cE873385dEc5CBbB6E45.pino.log](https://github.com/statechannels/monorepo/files/4625762/web3torrent.Thu.14.May.2020.02_42_14.GMT.f0abd39f4.0xC11C4a744747bEc56504cE873385dEc5CBbB6E45.pino.log)

**Comment by andrewgordstewart** [comment]
```
#! /bin/bash

# https://github.com/statechannels/monorepo/issues/1842

prettee () {
  tee $1.log | pino-pretty > $1.pretty.log
}

cat ~/Downloads/web3torrent.Thu,\ 14\ May\ 2020\ 02_42_14\ GMT.f0abd39f4.0xC11C4a744747bEc56504cE873385dEc5CBbB6E45.pino.log | prettee original 

# I think it's sorted, but ... just in case
cat original.log | jq -s -c 'sort_by(.time) | .[]' | prettee sorted

ERROR_TIME=1589424106834
cat sorted.log | jq  -c ". | select(.time <= $ERROR_TIME + 50 ) | select(.time >= $ERROR_TIME - 1000) " | prettee nearby

cat nearby.log | jq -c ". | select(.time == $ERROR_TIME) | .signedStates[] | {turnNum: .turnNum, outcome: .outcome}"  | head -n 2 | prettee duplicate-states
```

The duplicate states were the first two states listed, with turn number 4

```
    turnNum: {
      "_hex": "0x04"
    }
    outcome: {
      "type": "SimpleAllocation",
      "allocationItems": [
        {
          "destination": "0x000000000000000000000000fec44e15328B7d1d8885a8226b0858964358F1D6",
          "amount": {
            "_hex": "0x0000000000000000000000000000000000000000000000000000000000068000"
          }
        },
        {
          "destination": "0x000000000000000000000000760bf27cd45036a6C486802D30B5D90CfFBE31FE",
          "amount": {
            "_hex": "0x0000000000000000000000000000000000000000000000000000000000b09b00"
          }
        }
      ],
      "assetHolderAddress": "0x4ad3F07BEFDC54511449A1f553E36A653c82eA57"
    }
    participants: [
      {
        "participantId": "0x3daFD8E545E0F9763a038F96D99f6E1824690c3d",
        "signingAddress": "0x3daFD8E545E0F9763a038F96D99f6E1824690c3d",
        "destination": "0x000000000000000000000000d4Fa489Eacc52BA59438993f37Be9fcC20090E39"
      },
      {
        "participantId": "0xC11C4a744747bEc56504cE873385dEc5CBbB6E45",
        "signingAddress": "0xC11C4a744747bEc56504cE873385dEc5CBbB6E45",
        "destination": "0x000000000000000000000000760bf27cd45036a6C486802D30B5D90CfFBE31FE"
      }
    ]
    turnNum: {
      "_hex": "0x0000000000000000000000000000000000000000000000000000000000000004"
    }
    outcome: {
      "type": "SimpleAllocation",
      "assetHolderAddress": "0x4ad3F07BEFDC54511449A1f553E36A653c82eA57",
      "allocationItems": [
        {
          "destination": "0x000000000000000000000000d4Fa489Eacc52BA59438993f37Be9fcC20090E39",
          "amount": {
            "_hex": "0x0000000000000000000000000000000000000000000000000000000000000000"
          }
        },
        {
          "destination": "0x000000000000000000000000760bf27cd45036a6C486802D30B5D90CfFBE31FE",
          "amount": {
            "_hex": "0x0000000000000000000000000000000000000000000000000000000000b71b00"
          }
        }
      ]
    }
    participants: [
      {
        "participantId": "0x3daFD8E545E0F9763a038F96D99f6E1824690c3d",
        "signingAddress": "0x3daFD8E545E0F9763a038F96D99f6E1824690c3d",
        "destination": "0x000000000000000000000000d4Fa489Eacc52BA59438993f37Be9fcC20090E39"
      },
      {
        "participantId": "0xC11C4a744747bEc56504cE873385dEc5CBbB6E45",
        "signingAddress": "0xC11C4a744747bEc56504cE873385dEc5CBbB6E45",
        "destination": "0x000000000000000000000000760bf27cd45036a6C486802D30B5D90CfFBE31FE"
      }
    ]

```

One of the duplicataes contains a destination that's not a part of the channel
```
WRONG_DESTINATION=0x000000000000000000000000fec44e15328B7d1d8885a8226b0858964358F1D6

# Channel in question
ERROR_CHANNEL=0xf456d7157cb608216a2e6d94c5b27784a090273692b5a7bb46499804d4c0e319

cat nearby.log | grep $ERROR_CHANNEL | prettee errorChannel

grep $WRONG_DESTINATION errorChannel.log | prettee wrongDestination
```

I found this in wrongDestination.pretty.log
```
[1589424106716] INFO  (channel-provider): Sent message:
    message: {
      "jsonrpc": "2.0",
      "method": "UpdateChannel",
      "params": {
        "channelId": "0xf456d7157cb608216a2e6d94c5b27784a090273692b5a7bb46499804d4c0e319",
        "participants": [
          {
            "participantId": "0x080065862f469a229E73C41579AbdC3CF4cBEeD7",
            "signingAddress": "0x080065862f469a229E73C41579AbdC3CF4cBEeD7",
            "destination": "0x000000000000000000000000fec44e15328B7d1d8885a8226b0858964358F1D6"
          },
          {
            "participantId": "0xC11C4a744747bEc56504cE873385dEc5CBbB6E45",
            "signingAddress": "0xC11C4a744747bEc56504cE873385dEc5CBbB6E45",
            "destination": "0x000000000000000000000000760bf27cd45036a6C486802D30B5D90CfFBE31FE"
          }
        ],
        "allocations": [
          {
            "token": "0x0000000000000000000000000000000000000000",
            "allocationItems": [
              {
                "destination": "0x000000000000000000000000fec44e15328B7d1d8885a8226b0858964358F1D6",
                "amount": "0x0000000000000000000000000000000000000000000000000000000000068000"
              },
              {
                "destination": "0x000000000000000000000000760bf27cd45036a6C486802D30B5D90CfFBE31FE",
                "amount": "0x0000000000000000000000000000000000000000000000000000000000b09b00"
              }
            ]
          }
        ],
        "appData": "0x0000000000000000000000000000000000000000000000000000000000000000"
      },
      "id": 104
    }
```

Confirming in the store entry:

```
grep "Transaction error" nearby.log | jq -c ".store.channels[]" | grep "key\":\"$ERROR_CHANNEL" | prettee channelEntry

cat channelEntry.log | jq ".value.channelConstants.participants" | pbcopy
```

The "error channel" has these participants:
```
[
  {
    "participantId": "0x3daFD8E545E0F9763a038F96D99f6E1824690c3d",
    "signingAddress": "0x3daFD8E545E0F9763a038F96D99f6E1824690c3d",
    "destination": "0x000000000000000000000000d4Fa489Eacc52BA59438993f37Be9fcC20090E39"
  },
  {
    "participantId": "0xC11C4a744747bEc56504cE873385dEc5CBbB6E45",
    "signingAddress": "0xC11C4a744747bEc56504cE873385dEc5CBbB6E45",
    "destination": "0x000000000000000000000000760bf27cd45036a6C486802D30B5D90CfFBE31FE"
  }
]
```

So, is web3torrent asking to update the destination of the wrong channel?

**Comment by andrewgordstewart** [comment]
I think this is the problem: https://github.com/statechannels/monorepo/commit/fce88d4e20329355a7008b4d1b40fccf53628375

When making a payment for channel id `0xf00`, web3torrent was updating the channel that _last sent a payment receipt_, not `0xf00`

**Comment by andrewgordstewart** [comment]
When I run the `running` integration test on https://github.com/statechannels/monorepo/pull/1852/commits/26229644f92743d93912abc3a2a6e308cd0d01a7, then I get 
```
    Store signed multiple states for a single turn

      259 |       logger.error({entry: this.data()}, Errors.multipleSignedStates);
      260 | 
    > 261 |       throw Error(Errors.multipleSignedStates);
          |             ^
      262 |     }
      263 | 
      264 |     const {signedStates} = this;
```

**Comment by andrewgordstewart** [comment]
[error.log](https://github.com/statechannels/monorepo/files/4630085/error.log)

**Comment by snario** [comment]
Shouldn't the second one error with "not your turn"?

**Comment by andrewgordstewart** [comment]
Is that what the API should be?

What if
1. the app requests an update where they pay 10
2. the app changes their mind and wants to request a second update where they pay 12

In (1), how does the app know that it's their turn? Should the API expose that?

**Comment by snario** [comment]
Seems like:

- There should be an `override` flag which is documented but set to `false` by default which forces the wallet to sign and send a new state, even if they already signed a state before that. We could throw an error for now since we don't implement signing two states with the same turn number.

- A `ChannelState` has `turnNum` and `myIndex` right? `ChannelClient` could add a `isMyTurn` computed property based on that and it could be a "best practice" to only call `updateChannel` when that is true

**Comment by andrewgordstewart** [comment]
> Seems like:
> 
> * There should be an `override` flag which is documented but set to `false` by default which forces the wallet to sign and send a new state, even if they already signed a state before that. We could throw an error for now since we don't implement signing two states with the same turn number.

That sounds like a good thing to explore

> * A `ChannelState` has `turnNum` and `myIndex` right? `ChannelClient` could add a `isMyTurn` computed property based on that and it could be a "best practice" to only call `updateChannel` when that is true

I think we want some kind of behaviour a lot like this. But maybe the channel client should have pretty defensive behaviour that forces the application to have certain expectations about when and how they can update a channel.

When should the `updateChannel` promise resolve and reject to, and when? Should it be this?

(1) `updateChanel` rejects (quickly) if it is currently not your turn
(2) `updateChannel` resolves (quickly) when the client receives a response from the wallet that the updated state is in the store

In (1), how is it determined whether it's your turn? Should the client ask the wallet, or should it be according to the channel client's latest update?

Here is one proposal:

A. the client exposes a public `canUpdate` method, that rejects if there is an ongoing update, and otherwise resolves when I can requests an update
B. the client rejects `updateChannel` (immediately), if it is not the app's turn, according to the client's latest channel state
C. the client rejects `updateChannel` (immediately), if there is currently an `updateChannel` request
D. The client resolves `updateChannel`(quickly) when the client gets confirmation from the wallet that the channel is updated with the new outcome

So, A-C restrict the app to a single update per channel. (Perhaps this behaviour can change with an `override` flag.)

B & C don't have to make any queries to the wallet, since they can (and currently do) keep track of the latests channel state, and can compute `isMyTurn` based on `myIndex` and `turnNum`

## Note
These are all assuming a turn-taking application channel. That seems to explicitly rule out a "dynamic" null-app channel, where anyone can propose a move at any time. Do we want to rule that kind of channel out?

For the initial API, I think the answer is **yes**, we want to rule that kind of channel out.

**Comment by andrewgordstewart** [comment]
More thoughts:

A can mean: the client can reject immediately, without asking the wallet for any information (when there's an ongoing update. Or it can resolve eventually, when the wallet sends a `ChannelUpdated` event to the client.

D is the only time where the client needs to send an instruction to the wallet. I am not sure if the `updateChannel` promise should resolve precisely when the wallet **responds** to the `UpdateChannel`  JSON-RPC call? Or should it be when the subsequent `ChannelUpdated` message is received? @snario any thoughts?

**Comment by andrewgordstewart** [comment]
So @spersico's proposal is promise-chaining, something like the following:
1. The channel client manages a `pendingUpdate` for each channel. This value may be `undefined` or it may be a `Promise<ChannelUpdate>`. It is probably `private`.
1. `updateChannel(channelId)` resolves or rejects asynchronously, but quickly
   - it rejects if there is currently a `pendingUpdate` for that channel
   - it resolves with a `pendingUpdate` if there is no current `pendingUpdate` for that channel
2. `pendingUpdate` resolves with the current channel status, if the update is applied successfully
3. `pendingUpdate` rejects if the update cannot be applied
   - Maybe this is because the update is an invalid transition
   - Maybe this is because it is not the user's turn, and a timer has timed out

So, something like:

```
class ChannelClient {
  async updateChannel(params): Promise<ChannelResult> {
    const release = this.reservePendingUpdate(channelId);

    try {
      const ack = await this.provider.send({method: 'UpdateChannel', params});

      const pendingUpdate = new Promise((resolve, reject)=> {
        // wait for the channel update
        release();
        resolve("channel updated");
  
        // Or, if the wallet throws invalid transition
        release();
        reject("Invalid transition");
      })
  
      this.setPendingUpdate(channelId, pendingUpdate);
  
      return {ack, pendingUpdate}
    } catch (e) {
      release();
      return Promise.reject("Update request failed")
    }
  }
}
```

In other words, it's like submitting a transaction to the blockchain:
- `submit(tx)` returns a promise that resolves to a `pendingTransaction` promise
- `pendingTransaction` resolves when the transaction is mined, or rejects when something goes wrong

**Comment by andrewgordstewart** [comment]
This has some benefits:
1. `updateChannel` quickly resolves or quickly rejects (`ack` should resolve outside of xstate-wallet workflows, for instance, just a quick "I got the update request, and am working on it"). In particular, no network calls are needed for `updateChannel` to resolve.
2. `updateChannel` is a valid call at any time. It's ok to ask to update a channel when it's not your turn
3. Asking the wallet to make concurrent updates to a single channel is too complex. Whether the 2nd update succeeds depends on an intermediate state that, in general, the application does not know about.
4. `pendingUpdate` is a developer-friendly object. Applications can choose to block on it before continuing, or if they don't care, they can ignore it.

2 & 3 are a bit related -- with 2, the app either
A. needs to predict whether or not the proposed update will form a valid transition
B. needs to handle an invalid transition

but I think the app should handle B anyway, so (2) seems like a good property.

3 seems to have much more limited use cases, and apps can implement their own update queue if they wanted to.

Within the wallet, the application workflow can:
1. wait until it's my turn
2. try to apply the update
3. if it succeeds, cause the `pendingUpdate` to resolve by sending a message back
4. if it fails, cause the `pendingUpdate` to reject by sending a message back
5. Maybe, trigger a challenge workflow after a timeout period. Or pop up a modal ("It's still not your turn. Do you want to take action X or Y?")


1-4 seems like level 0. 5 seems like level 2, though "reject after a timeout elapses" should probably be level 0 behaviour.

**Comment by andrewgordstewart** [comment]
Pausing this until we come to agreement on the expected behaviour of `updateChannel`.

Once that's sorted out, there's two clear sets of work in the `channel-client` and `xstate-wallet` packages.

**Comment by snario** [comment]
Anything left to be decided @andrewgordstewart ?

**Comment by andrewgordstewart** [comment]
> Anything left to be decided @andrewgordstewart ?

https://github.com/statechannels/monorepo/issues/1853#issuecomment-630883292

Are these draft proposals considered to be accepted? We should probably discuss at standup.
